kernel/src/engine/default/storage.rs

```diff
@@ -1,38 +1,141 @@
-#[cfg(feature = "cloud")]
-use hdfs_native_object_store::HdfsObjectStore;
 use object_store::parse_url_opts as parse_url_opts_object_store;
 use object_store::path::Path;
 use object_store::{Error, ObjectStore};
 use url::Url;
 
+use std::collections::HashMap;
+use std::sync::{Arc, OnceLock, RwLock};
+
+/// Alias for convenience
+type ClosureReturn = Result<(Box<dyn ObjectStore>, Path), Error>;
+/// This type alias makes it easier to reference the handler closure(s)
+///
+/// It uses a HashMap<String, String> which _must_ be converted in our [parse_url_opts] because we
+/// cannot use generics in this scenario.
+type HandlerClosure = Arc<dyn Fn(&Url, HashMap<String, String>) -> ClosureReturn + Send + Sync>;
+/// hashmap containing scheme => handler fn mappings to allow consumers of delta-kernel-rs provide
+/// their own url opts parsers for different scemes
+type Handlers = HashMap<String, HandlerClosure>;
+/// The URL_REGISTRY contains the custom URL scheme handlers that will parse URL options
+static URL_REGISTRY: OnceLock<RwLock<Handlers>> = OnceLock::new();
+
+/// Return to fully initialied URL handlers to allow extensible provisioning of [object_store] URL
+/// handlers
+fn get_url_handlers<'life>() -> &'life RwLock<Handlers> {
+    URL_REGISTRY.get_or_init(|| RwLock::new(HashMap::default()))
+}
+
+/// Insert a new URL handler for [parse_url_opts] with the given `scheme`. This allows users to
+/// provide their own custom URL handler to plug new [object_store::ObjectStore] instances into
+/// delta-kernel
+pub fn insert_url_handler(
```

Comment:
I wonder if we should add this to the default `StorageHandler`? Would it be easier to use there instead of just as a standalone? (could still stay a static function but just be invoked like `StorageHandler::insert_url_handler`?

Reply:
We don't currently expose the storage handler as an arg in `new`. So we'd have to modify the trait in the core kernel to include this so that you could `get_storage_handler` and then `insert_url_handler`, which seems gross.

I'd say like this is okay for now, and we can refactor to pass the StorageHandler as part of the call to `new` in the future if we feel it's better.